Create a page for OpenFOAM (#17)
ACRC/HPC-community-docs: source/engineering/OpenFOAM/single-node.md

## Job Script Template

Here is an example job submission script for a single-node OpenFOAM job:

```{code-block} bash
---
linenos: true
---
#!/bin/bash
## Submission script for Cluster
#SBATCH --job-name=my_job
#SBATCH --time=0-01:00:00 # hh:mm:ss
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=1

## system error message output file
## leave %j as it's being replaced by JOB ID number
#SBATCH -e my_job_%j.err

## system message output file
#SBATCH -o my_job_%j.out

# record some potentially useful details about the job: 
echo Running on host $(hostname) 
echo Time is $(date) 
echo Directory is $(pwd) 
echo Slurm job ID is ${SLURM_JOBID} 
echo This jobs runs on the following machines: 
echo ${SLURM_JOB_NODELIST} 
printf "\n\n\n\n" 

# Load modules required for runtime e.g.
module load apps/openfoam/6

# Run the solver. Take pisoFoam for example:

pisoFoam

echo End Time is $(date) 
echo "Done pisoFoam finish"
printf "\n\n"

```

```{note}
For further information on the structure and syntax of Slurm job scripts, see the [ACRC HPC documentation pages on job types](https://www.acrc.bris.ac.uk/protected/hpc-docs/job_types/index.html).
```

### How to use

> Save the script file in the [case folder]([case folder](https://www.openfoam.com/documentation/user-guide/2-openfoam-cases), `cd` into the case folder, and submit the script to Slurm.

Reviewer: ```suggestion
Save the script file in the [case folder](https://www.openfoam.com/documentation/user-guide/2-openfoam-cases), `cd` into the case folder, and submit the script to Slurm.
```

---
title: "Single-Node Jobs"
---

Small OpenFOAM jobs can be run on a single node.

## Job Script Template

Here is an example job submission script for a single-node OpenFOAM job:

```{code-block} bash
---
linenos: true
---
#!/bin/bash
## Submission script for Cluster
#SBATCH --job-name=my_job
#SBATCH --time=0-01:00:00 # hh:mm:ss
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=1

> #SBATCH --partition=cpu

Reviewer: ```suggestion
```

There is no need to specify the `cpu` partition, as this is the default on BC4.

## system error message output file
## leave %j as it's being replaced by JOB ID number

> #SBATCH -e aai_LD_%j.err

Reviewer: What does `aai_LD` in the filename mean here? Maybe it would be simpler to rename to `my_job_%j.err` and `my_job_%j.out`?
```suggestion
#SBATCH -e my_job_%j.err
```

## system message output file

> #SBATCH -o aai_LD_%j.out

Reviewer: ```suggestion
#SBATCH -o my_job_%j.out
```

# record some potentially useful details about the job: 
echo Running on host $(hostname) 
echo Time is $(date) 
echo Directory is $(pwd) 
echo Slurm job ID is ${SLURM_JOBID} 
echo This jobs runs on the following machines: 
echo ${SLURM_JOB_NODELIST} 
printf "\n\n\n\n" 

# Load modules required for runtime e.g.
module load apps/openfoam/6

> #export I_MPI_PROCESS_MANAGER=mpd

Reviewer: Is this commented line needed? If not, I suggest removing it.

```suggestion
```

Author: Let me try a case without this code. We tried many scripts for the multi-processes case, and finally this one works. But you are right, this commented line doesn't work.

> 

Reviewer: Should the script cd to the case folder before running pisoFoam?

Author: I recommend users save the script file in the case folder and submit the job, so I didn't add cd command in the script.

Reviewer: Okay, do you mean that you need to submit the script from the case folder? 

If so, I suggest making this clear in the "How to use section".

It would be possible to save the script in the case folder, but submit it from somewhere else, e.g. submitting from the root of the home directory

```ShellSession
$ cd ${HOME}
$ sbatch case-folder/submit.sh
```

Author: This also works. I think submitting the script in the case folder directly is more convenient.

# Run the solver. Take pisoFoam for example:

pisoFoam

echo End Time is $(date) 

> echo "Done pimpleFoam finish"

Reviewer: ```suggestion
echo "Done pisoFoam finish"
```

printf "\n\n"

```

```{note}

> Further explanation of the script could be found in: https://www.acrc.bris.ac.uk/protected/hpc-docs/job_types/index.html

Reviewer: ```suggestion
For further information on the structure and syntax of Slurm job scripts, see the [ACRC HPC documentation pages on job types](https://www.acrc.bris.ac.uk/protected/hpc-docs/job_types/index.html).
```

```

### How to use

> Save the script file in the case folder, and submit it to Slurm.

Reviewer: ```suggestion
Save the script file in the case folder, `cd` into the case folder, and submit the script to Slurm.
```
